File: cores/typescript/src/region.ts
// TV system / region (port of cores/c/src/region.c).

export enum Region {
    Ntsc = 0,
    Pal = 1,
    Dendy = 2,
}

export interface ApuFrameThreshold {
    threshold: number;
    quarter: boolean;
    half: boolean;
}

const Ntsc4Step = new Uint32Array([7457, 14913, 22371, 29828]);
const Pal4Step = new Uint32Array([8314, 16627, 24941, 33255]);
const Ntsc5Step = new Uint32Array([7457, 14913, 22371, 37281]);
const Pal5Step = new Uint32Array([8314, 16627, 24941, 41568]);
const QuarterFlags = [true, true, true, true];
const HalfFlags = [false, true, false, true];

export function scanlinesPerFrame(r: number): number {
    if (r === Region.Ntsc) return 262;
    if (r === Region.Pal) return 312;
    return 312; // Dendy
}

export function scanlinePrerender(r: number): number {
    if (r === Region.Ntsc) return 261;
    if (r === Region.Pal) return 311;
    return 311; // Dendy
}

export function isPalPalette(r: number): boolean {
    return r === Region.Pal;
}

export function cpuClockHz(r: number): number {
    if (r === Region.Pal) return 1662607.0;
    return 1789773.0;
}

export function cpuCyclesPerSample(r: number): number {
    return cpuClockHz(r) / 44100.0;
}

export function apu4StepThresholds(r: number, outs: ApuFrameThreshold[]): void {
    const t = (r === Region.Pal) ? Pal4Step : Ntsc4Step;
    for (let i = 0; i < 4; ++i) {
        outs[i].threshold = t[i];
        outs[i].quarter = QuarterFlags[i];
        outs[i].half = HalfFlags[i];
    }
}

export function apu5StepThresholds(r: number, outs: ApuFrameThreshold[]): void {
    const t = (r === Region.Pal) ? Pal5Step : Ntsc5Step;
    for (let i = 0; i < 4; ++i) {
        outs[i].threshold = t[i];
        outs[i].quarter = QuarterFlags[i];
        outs[i].half = HalfFlags[i];
    }
}

export function apuResetAt(r: number, mode5step: boolean): number {
    if (r === Region.Pal) return mode5step ? 41570 : 33257;
    return mode5step ? 37282 : 29830;
}

export function apu4StepIrqThreshold(r: number): number {
    if (r === Region.Pal) return 33255;
    return 29828;
}

File: cores/csharp/src/Region.cs
using System.Runtime.CompilerServices;

namespace NesCore;

// TV system / region (src/region.rs Region).
public enum Region : byte
{
    Ntsc = 0,
    Pal = 1,
    Dendy = 2,
}

// APU frame-counter threshold: cycle position + quarter/half flags.
public struct ApuFrameThreshold
{
    public uint Threshold;
    public bool Quarter;
    public bool Half;
}

// Region timing accessors (port of src/region.rs).
public static class RegionTiming
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static ushort ScanlinesPerFrame(Region r) => r switch
    {
        Region.Ntsc => 262,
        Region.Pal => 312,
        Region.Dendy => 312,
        _ => 262,
    };

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static ushort ScanlinePrerender(Region r) => r switch
    {
        Region.Ntsc => 261,
        Region.Pal => 311,
        Region.Dendy => 311,
        _ => 261,
    };

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsPalPalette(Region r) => r == Region.Pal;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static float CpuClockHz(Region r) => r switch
    {
        Region.Ntsc => 1789773.0f,
        Region.Pal => 1662607.0f,
        Region.Dendy => 1789773.0f,
        _ => 1789773.0f,
    };

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static float CpuCyclesPerSample(Region r) => CpuClockHz(r) / 44100.0f;

    // Pre-allocated threshold tables (avoid per-call allocation in hot path).
    private static readonly uint[] Ntsc4Step = { 7457u, 14913u, 22371u, 29828u };
    private static readonly uint[] Pal4Step = { 8314u, 16627u, 24941u, 33255u };
    private static readonly uint[] Ntsc5Step = { 7457u, 14913u, 22371u, 37281u };
    private static readonly uint[] Pal5Step = { 8314u, 16627u, 24941u, 41568u };
    private static readonly bool[] QuarterFlags = { true, true, true, true };
    private static readonly bool[] HalfFlags = { false, true, false, true };

    public static void Apu4StepThresholds(Region r, ApuFrameThreshold[] outs)
    {
        uint[] t = (r == Region.Pal) ? Pal4Step : Ntsc4Step;
        for (int i = 0; i < 4; ++i)
        {
            outs[i].Threshold = t[i];
            outs[i].Quarter = QuarterFlags[i];
            outs[i].Half = HalfFlags[i];
        }
    }

    public static void Apu5StepThresholds(Region r, ApuFrameThreshold[] outs)
    {
        uint[] t = (r == Region.Pal) ? Pal5Step : Ntsc5Step;
        for (int i = 0; i < 4; ++i)
        {
            outs[i].Threshold = t[i];
            outs[i].Quarter = QuarterFlags[i];
            outs[i].Half = HalfFlags[i];
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static uint ApuResetAt(Region r, bool mode5step) => r switch
    {
        Region.Ntsc or Region.Dendy => mode5step ? 37282u : 29830u,
        Region.Pal => mode5step ? 41570u : 33257u,
        _ => mode5step ? 37282u : 29830u,
    };

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static uint Apu4StepIrqThreshold(Region r) => r switch
    {
        Region.Ntsc or Region.Dendy => 29828u,
        Region.Pal => 33255u,
        _ => 29828u,
    };
}
